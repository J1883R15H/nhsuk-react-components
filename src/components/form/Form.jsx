import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import FormContext from './FormContext';
import { Heading } from '../typography';

class Form extends Component {
  static propTypes = {
    className: PropTypes.string,
    onChange: PropTypes.func,
    onSubmit: PropTypes.func,
    children: PropTypes.node.isRequired,
    onErrorChange: PropTypes.func
  };

  static defaultProps = {
    className: '',
    onChange: () => {},
    onSubmit: () => {},
    onErrorChange: () => {}
  };

  constructor(props) {
    super(props);
    this.state = {
      formData: {},
      erroredComponents: {}
    };
  }

  registerComponent = (name, value) => {
    this.setState(() => {
      const { formData } = this.state;
      formData[name] = value;
      return { formData };
    });
  };

  updateFormState = (name, value) => {
    const { onChange } = this.props;
    this.setState(() => {
      const { formData } = this.state;
      formData[name] = value;
      if (onChange) onChange(formData);
      return { formData };
    });
  };

  onSubmit = e => {
    const { onSubmit } = this.props;
    const { formData } = this.state;
    if (onSubmit) onSubmit(e, formData);
  };

  passBackError = (name, error = true, errorDetails) => {
    const { erroredComponents } = this.state;
    const original = erroredComponents;
    if (error) {
      erroredComponents[name] = errorDetails;
    } else if (Object.keys(erroredComponents).includes(name)) {
      delete erroredComponents[name];
    }

    if (original !== erroredComponents) {
      this.setState({ erroredComponents }, () => {
        const { onErrorChange } = this.props;
        onErrorChange(erroredComponents);
      });
    }
  };

  render() {
    const {
      className,
      children,
      action,
      method,
      title,
      titleSize,
      ...rest
    } = this.props;
    const { erroredComponents } = this.state;
    const contextValue = {
      registerComponent: this.registerComponent,
      updateFormState: this.updateFormState,
      passBackError: this.passBackError
    };
    return (
      <form action={action} method={method} onSubmit={this.onSubmit}>
        <div
          className={classNames(
            'nhsuk-form-group',
            { 'nhsuk-form-group--error': erroredComponents.length > 0 },
            className
          )}
          {...rest}
        >
          {title ? <Heading size={titleSize}>{title}</Heading> : null}
          <FormContext.Provider value={contextValue}>
            {children}
          </FormContext.Provider>
        </div>
      </form>
    );
  }
}

export default Form;
