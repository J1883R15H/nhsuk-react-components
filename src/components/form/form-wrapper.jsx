import React from 'react';
import PropTypes from 'prop-types';
import stylePropType from 'react-style-proptype';
import './_form.scss';
import Heading from '../typography/heading';

class Form extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      formData: {}
    };
    this.valueCallback = this.valueCallback.bind(this);
    this.registerInitialValue = this.registerInitialValue.bind(this);
  }

  registerInitialValue(key, value) {
    this.setState(previousState => {
      const previousFormData = previousState.formData;
      return { formData: { ...previousFormData, [key]: value } };
    });
  }

  valueCallback(key, value) {
    const { onChange } = this.props;
    this.setState(previousState => {
      const previousFormData = previousState.formData;
      onChange({ formData: { ...previousFormData, [key]: value } });
      return { formData: { ...previousFormData, [key]: value } };
    });
  }

  render() {
    const {
      submitAction,
      submitMethod,
      children,
      style,
      className,
      title,
      titleSize,
      elementMargin,
      onSubmit
    } = this.props;
    const { formData } = this.state;
    let errorInChild = false;
    const modifiedChildren = React.Children.map(children, child => {
      const { props } = child;
      if (props) {
        if (props.error) {
          errorInChild = true;
        }
        const formElements = [
          'Checkboxes',
          'DateInput',
          'Input',
          'InputBlock',
          'Radios',
          'Select',
          'Textarea'
        ];
        if (formElements.includes(child.type.name)) {
          return React.cloneElement(child, {
            valueCallback: this.valueCallback,
            style: { ...child.props.style, marginBottom: elementMargin },
            registerInitialValue: this.registerInitialValue
          });
        }
        return child;
      }
      return child;
    });
    return (
      <form
        action={submitAction}
        method={submitMethod}
        onSubmit={e => onSubmit(e, formData)}
      >
        {title ? <Heading size={titleSize}>{title}</Heading> : null}
        <div
          className={`nhsuk-form-group ${
            errorInChild ? 'nhsuk-form-group--error' : null
          } ${className}`}
          style={style}
        >
          {modifiedChildren}
        </div>
      </form>
    );
  }
}

Form.propTypes = {
  submitAction: PropTypes.string,
  submitMethod: PropTypes.string,
  children: PropTypes.node.isRequired,
  onSubmit: PropTypes.func,
  className: PropTypes.string,
  style: stylePropType,
  elementMargin: PropTypes.number,
  onChange: PropTypes.func
};

Form.defaultProps = {
  submitAction: '',
  submitMethod: 'GET',
  className: '',
  style: {},
  elementMargin: 0,
  onSubmit: () => {},
  onChange: () => {}
};

export default Form;
