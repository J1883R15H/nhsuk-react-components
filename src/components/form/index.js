import Form from './form-wrapper';
import Checkboxes from './components/checkboxes';
import DateInput from './components/date-input';
import Input from './components/input';
import Radios from './components/radios';
import Select from './components/select';
import Textarea from './components/textarea';
import InputBlock from './components/input-block';

export {
  Form,
  Checkboxes,
  DateInput,
  Input,
  Radios,
  Select,
  Textarea,
  InputBlock
};
