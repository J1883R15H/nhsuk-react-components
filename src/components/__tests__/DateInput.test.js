import React from 'react';
import { mount, shallow } from 'enzyme';
import { DateInput } from '../form';
import Hint from '../hint';
import Label from '../label';
import ErrorMessage from '../error-message';

describe('DateInput', () => {
  it('all subelements have valid defaultProps', () => {
    expect(DateInput.Day.defaultProps._handleInput()).toBe(undefined);
    expect(DateInput.Month.defaultProps._handleInput()).toBe(undefined);
    expect(DateInput.Year.defaultProps._handleInput()).toBe(undefined);
    expect(DateInput.Month.defaultProps._registerComponent()).toBe(undefined);
    expect(DateInput.Year.defaultProps._registerComponent()).toBe(undefined);
  });
  describe('Day', () => {
    it('contains elements', () => {
      const wrapper = shallow(<DateInput.Day _name="name" />);
      expect(
        wrapper.containsMatchingElement(
          <input
            className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
            id="name-day"
            name="name-day"
            aria-label="name-day input"
            type="number"
            autoComplete=""
            value=""
            pattern="[0-9]*"
            required={false}
          />
        )
      ).toBeTruthy();
      wrapper.unmount();
    });
    it('passes back input', () => {
      const handleInput = jest.fn();
      const wrapper = shallow(
        <DateInput.Day _name="name" _handleInput={handleInput} />
      );
      const wrapperForAnonFunc = shallow(
        <DateInput.Day _name="name" _handleInput={handleInput} />
      );
      wrapper.find('input').simulate('change', { target: { value: '10' } });
      wrapperForAnonFunc
        .find('input')
        .simulate('change', { target: { value: '10' } });
      expect(handleInput).toHaveBeenCalled();
      expect(handleInput.mock.calls[0][0]).toBe('day');
      expect(handleInput.mock.calls[0][1]).toBe('10');
      wrapper.unmount();
    });
    it('renders with error', () => {
      const wrapper = shallow(<DateInput.Day _name="name" error="error" />);
      expect(wrapper.find('input').props().className).toBe(
        'nhsuk-input nhsuk-date-input__input nhsuk-input--width-2 nhsuk-input--error'
      );
      wrapper.unmount();
    });
  });

  describe('Month', () => {
    it('contains elements', () => {
      const wrapper = shallow(<DateInput.Month _name="name" />);
      expect(
        wrapper.containsMatchingElement(
          <input
            className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
            id="name-month"
            name="name-month"
            aria-label="name-month input"
            type="number"
            autoComplete=""
            value=""
            pattern="[0-9]*"
            required={false}
          />
        )
      ).toBeTruthy();
      wrapper.unmount();
    });
    it('passes back input', () => {
      const handleInput = jest.fn();
      const wrapper = shallow(
        <DateInput.Month _name="name" _handleInput={handleInput} />
      );
      const wrapperForAnonFunc = shallow(<DateInput.Month _name="name" />);
      wrapper.find('input').simulate('change', { target: { value: '10' } });
      wrapperForAnonFunc
        .find('input')
        .simulate('change', { target: { value: '10' } });
      expect(handleInput).toHaveBeenCalled();
      expect(handleInput.mock.calls[0][0]).toBe('month');
      expect(handleInput.mock.calls[0][1]).toBe('10');
      wrapper.unmount();
    });
    it('registers ref', () => {
      const registerComponent = jest.fn();
      const wrapper = mount(
        <DateInput.Month _name="name" _registerComponent={registerComponent} />
      );
      expect(registerComponent).toHaveBeenCalled();
      expect(registerComponent.mock.calls[0][0]).toBe('month');
      wrapper.unmount();
    });
    it('renders with error', () => {
      const wrapper = shallow(<DateInput.Month _name="name" error="error" />);
      expect(wrapper.find('input').props().className).toBe(
        'nhsuk-input nhsuk-date-input__input nhsuk-input--width-2 nhsuk-input--error'
      );
      wrapper.unmount();
    });
  });
  describe('Year', () => {
    it('contains elements', () => {
      const wrapper = shallow(<DateInput.Year _name="name" />);
      expect(
        wrapper.containsMatchingElement(
          <input
            aria-label="name-year input"
            autoComplete=""
            className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
            id="name-year"
            name="name-year"
            pattern="[0-9]*"
            type="number"
            value=""
          />
        )
      ).toBeTruthy();
      wrapper.unmount();
    });
    it('passes back input', () => {
      const handleInput = jest.fn();
      const wrapper = shallow(
        <DateInput.Year _name="name" _handleInput={handleInput} />
      );
      const wrapperForAnonFunc = shallow(<DateInput.Year _name="name" />);
      wrapper.find('input').simulate('change', { target: { value: '10' } });
      wrapperForAnonFunc
        .find('input')
        .simulate('change', { target: { value: '10' } });
      expect(handleInput).toHaveBeenCalled();
      expect(handleInput.mock.calls[0][0]).toBe('year');
      expect(handleInput.mock.calls[0][1]).toBe('10');
      wrapper.unmount();
    });
    it('registers ref', () => {
      const registerComponent = jest.fn();
      const wrapper = mount(
        <DateInput.Year _name="name" _registerComponent={registerComponent} />
      );
      expect(registerComponent).toHaveBeenCalled();
      expect(registerComponent.mock.calls[0][0]).toBe('year');
      wrapper.unmount();
    });
    it('renders with error', () => {
      const wrapper = shallow(<DateInput.Year _name="name" error="error" />);
      expect(wrapper.find('input').props().className).toBe(
        'nhsuk-input nhsuk-date-input__input nhsuk-input--width-4 nhsuk-input--error'
      );
      wrapper.unmount();
    });
  });
  describe('Wrapper', () => {
    it('renders without individual components', () => {
      const wrapper = mount(<DateInput name="dateInput" />);
      expect(wrapper).toMatchSnapshot();
      wrapper.unmount();
    });
    it("doesn't call handleAutoFocus", () => {
      const wrapper = mount(<DateInput name="dateInput" />);
      const spy = jest.spyOn(wrapper.instance(), 'handleAutoFocus');
      wrapper
        .find('#dateInput-day')
        .simulate('change', { target: { value: '10' } });
      expect(spy).not.toHaveBeenCalled();
      wrapper.unmount();
    });

    it('matches snapshot', () => {
      const multiMount = mount(
        <DateInput name="dateInput">
          <DateInput.Day />
          <DateInput.Month />
          <DateInput.Year />
        </DateInput>
      );
      expect(multiMount).toMatchSnapshot();
      multiMount.unmount();
    });
    it('calls back values', () => {
      const valueCallback = jest.fn();
      const wrapper = mount(
        <DateInput valueCallback={valueCallback} autoFocus name="dateInput" />
      );

      expect(valueCallback).toHaveBeenCalledTimes(0);
      wrapper
        .find('#dateInput-day')
        .simulate('change', { target: { value: '10' } });
      expect(valueCallback).toHaveBeenCalledTimes(1);
      expect(valueCallback.mock.calls[0]).toEqual([
        'dateInput',
        {
          day: '10',
          month: '',
          year: ''
        }
      ]);
      wrapper
        .find('#dateInput-month')
        .simulate('change', { target: { value: '10' } });
      expect(valueCallback).toHaveBeenCalledTimes(2);
      expect(valueCallback.mock.calls[1]).toEqual([
        'dateInput',
        {
          day: '10',
          month: '10',
          year: ''
        }
      ]);
      wrapper
        .find('#dateInput-year')
        .simulate('change', { target: { value: '2000' } });
      expect(valueCallback).toHaveBeenCalledTimes(3);
      expect(valueCallback.mock.calls[2]).toEqual([
        'dateInput',
        {
          day: '10',
          month: '10',
          year: '2000'
        }
      ]);
      wrapper.unmount();
    });
  });
  it("doesn't inject props into other children", () => {
    const wrapper = mount(
      <DateInput name="dateInput">
        <h1>h1</h1>
      </DateInput>
    );
    expect(wrapper.containsMatchingElement(<h1>h1</h1>)).toBeTruthy();
    wrapper.unmount();
  });

  it('renders with error, label and hint', () => {
    const wrapper = shallow(
      <DateInput name="dateInput" hint="hint" label="Label" error="error" />
    );
    expect(wrapper.containsMatchingElement(<Hint>hint</Hint>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<Label>Label</Label>)).toBeTruthy();
    expect(
      wrapper.containsMatchingElement(<ErrorMessage>error</ErrorMessage>)
    ).toBeTruthy();
    wrapper.unmount();
  });
  it('registers component', () => {
    const wrapper = shallow(
      <DateInput name="dateInput" hint="hint" error="error" />
    );
    wrapper.instance().registerComponent('month', 'ref');
    wrapper.instance().registerComponent('year', 'ref');
    wrapper.instance().registerComponent('year1', 'ref');
    expect(wrapper.instance().monthInput).toBe('ref');
    expect(wrapper.instance().yearInput).toBe('ref');
    wrapper.unmount();
  });
});
