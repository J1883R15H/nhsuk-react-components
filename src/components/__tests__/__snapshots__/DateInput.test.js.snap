// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput Wrapper matches snapshot 1`] = `
<DateInput
  asPageHeading={false}
  autoFocus={false}
  className=""
  describedBy=""
  error=""
  hint=""
  id=""
  name="dateInput"
  registerInitialValue={[Function]}
  required={false}
  style={Object {}}
  title=""
  valueCallback={[Function]}
>
  <Fieldset
    asPageHeading={false}
    className=""
    describedBy=""
    style={Object {}}
    title=""
  >
    <fieldset
      aria-describedby=""
      className="nhsuk-fieldset"
      style={Object {}}
    >
      <legend
        className="nhsuk-fieldset__legend"
      />
      <div
        className="nhsuk-date-input"
        id=""
      >
        <Day
          _day=""
          _handleInput={[Function]}
          _name="dateInput"
          autoComplete=""
          className=""
          error={false}
          key=".0"
          label="Day"
          required={false}
          style={Object {}}
        >
          <div
            className="nhsuk-date-input__item"
            style={Object {}}
          >
            <div
              className="nhsuk-form-group"
            >
              <Label
                className=""
                htmlFor="dateInput-day"
                htmlForId=""
                style={Object {}}
                type=""
              >
                <label
                  className="nhsuk-label"
                  htmlFor=""
                  style={Object {}}
                >
                  Day
                </label>
              </Label>
              <input
                aria-label="dateInput-day input"
                autoComplete=""
                className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                id="dateInput-day"
                name="dateInput-day"
                onChange={[Function]}
                pattern="[0-9]*"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </Day>
        <Month
          _handleInput={[Function]}
          _month=""
          _name="dateInput"
          _registerComponent={[Function]}
          autoComplete=""
          className=""
          error={false}
          key=".1"
          label="Month"
          required={false}
          style={Object {}}
        >
          <div
            className="nhsuk-date-input__item"
            style={Object {}}
          >
            <div
              className="nhsuk-form-group"
            >
              <Label
                className=""
                htmlFor="dateInput-day"
                htmlForId=""
                style={Object {}}
                type=""
              >
                <label
                  className="nhsuk-label"
                  htmlFor=""
                  style={Object {}}
                >
                  Month
                </label>
              </Label>
              <input
                aria-label="dateInput-month input"
                autoComplete=""
                className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                id="dateInput-month"
                name="dateInput-month"
                onChange={[Function]}
                pattern="[0-9]*"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </Month>
        <Year
          _handleInput={[Function]}
          _name="dateInput"
          _registerComponent={[Function]}
          _year=""
          autoComplete=""
          className=""
          error={false}
          key=".2"
          label="Year"
          required={false}
          style={Object {}}
        >
          <div
            className="nhsuk-date-input__item"
            style={Object {}}
          >
            <div
              className="nhsuk-form-group"
            >
              <Label
                className=""
                htmlFor="dateInput-day"
                htmlForId=""
                style={Object {}}
                type=""
              >
                <label
                  className="nhsuk-label"
                  htmlFor=""
                  style={Object {}}
                >
                  Year
                </label>
              </Label>
              <input
                aria-label="dateInput-year input"
                autoComplete=""
                className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                id="dateInput-year"
                name="dateInput-year"
                onChange={[Function]}
                pattern="[0-9]*"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </Year>
      </div>
    </fieldset>
  </Fieldset>
</DateInput>
`;

exports[`DateInput Wrapper renders without individual components 1`] = `
<DateInput
  asPageHeading={false}
  autoFocus={false}
  className=""
  describedBy=""
  error=""
  hint=""
  id=""
  name="dateInput"
  registerInitialValue={[Function]}
  required={false}
  style={Object {}}
  title=""
  valueCallback={[Function]}
>
  <Fieldset
    asPageHeading={false}
    className=""
    describedBy=""
    style={Object {}}
    title=""
  >
    <fieldset
      aria-describedby=""
      className="nhsuk-fieldset"
      style={Object {}}
    >
      <legend
        className="nhsuk-fieldset__legend"
      />
      <div
        className="nhsuk-date-input"
        id=""
      >
        <Day
          _day=""
          _handleInput={[Function]}
          _name="dateInput"
          autoComplete=""
          className=""
          error={false}
          label="Day"
          required={false}
          style={Object {}}
        >
          <div
            className="nhsuk-date-input__item"
            style={Object {}}
          >
            <div
              className="nhsuk-form-group"
            >
              <Label
                className=""
                htmlFor="dateInput-day"
                htmlForId=""
                style={Object {}}
                type=""
              >
                <label
                  className="nhsuk-label"
                  htmlFor=""
                  style={Object {}}
                >
                  Day
                </label>
              </Label>
              <input
                aria-label="dateInput-day input"
                autoComplete=""
                className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                id="dateInput-day"
                name="dateInput-day"
                onChange={[Function]}
                pattern="[0-9]*"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </Day>
        <Month
          _handleInput={[Function]}
          _month=""
          _name="dateInput"
          _registerComponent={[Function]}
          autoComplete=""
          className=""
          error={false}
          label="Month"
          required={false}
          style={Object {}}
        >
          <div
            className="nhsuk-date-input__item"
            style={Object {}}
          >
            <div
              className="nhsuk-form-group"
            >
              <Label
                className=""
                htmlFor="dateInput-day"
                htmlForId=""
                style={Object {}}
                type=""
              >
                <label
                  className="nhsuk-label"
                  htmlFor=""
                  style={Object {}}
                >
                  Month
                </label>
              </Label>
              <input
                aria-label="dateInput-month input"
                autoComplete=""
                className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-2"
                id="dateInput-month"
                name="dateInput-month"
                onChange={[Function]}
                pattern="[0-9]*"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </Month>
        <Year
          _handleInput={[Function]}
          _name="dateInput"
          _registerComponent={[Function]}
          _year=""
          autoComplete=""
          className=""
          error={false}
          label="Year"
          required={false}
          style={Object {}}
        >
          <div
            className="nhsuk-date-input__item"
            style={Object {}}
          >
            <div
              className="nhsuk-form-group"
            >
              <Label
                className=""
                htmlFor="dateInput-day"
                htmlForId=""
                style={Object {}}
                type=""
              >
                <label
                  className="nhsuk-label"
                  htmlFor=""
                  style={Object {}}
                >
                  Year
                </label>
              </Label>
              <input
                aria-label="dateInput-year input"
                autoComplete=""
                className="nhsuk-input nhsuk-date-input__input nhsuk-input--width-4"
                id="dateInput-year"
                name="dateInput-year"
                onChange={[Function]}
                pattern="[0-9]*"
                required={false}
                type="number"
                value=""
              />
            </div>
          </div>
        </Year>
      </div>
    </fieldset>
  </Fieldset>
</DateInput>
`;
